### PR TITLE
SALTO-1730 - Salesforce: Refactor formula utils

### PR DESCRIPTION
**Pull request by tomsellek**
Refactor the code we copied from Forcemula

---
~This PR builds on top of https://github.com/salto-io/salto/pull/4005, so the first two commits should be ignored.~
I put each individual refactoring step into its own commit, which should hopefully make reviewing easier.

---
_Release Notes_: 
N/A

---
_User Notifications_: 
N/A

### CODE REVIEW

**Review by amitgoldarb — APPROVED** [review]
I added a few small comments. Approving because this only makes things better
Overall I'm wondering if we can remove some of the intermediary types like IdentifierType here, this code "feels" a bit out of place and not styled like the rest of the adapter. Not sure if the refactor would be entirely worth the gain though.

File: packages/salesforce-adapter/src/filters/formula_deps.ts [hunk 1]
```diff
@@ -34,18 +34,18 @@ const log = logger(module)
 const { awu } = collections.asynciterable
 
 const identifierTypeToElementName = (identifierInfo: FormulaIdentifierInfo): string[] => {
-  if (identifierInfo.type === IdentifierType.CUSTOM_LABEL) {
+  if (identifierInfo.type === 'customLabel') {
```
[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/src/filters/formula_deps.ts:37]: This isn't a string but rather an `IdentifierType`

File: packages/salesforce-adapter/src/filters/formula_deps.ts [hunk 2]
```diff
@@ -54,15 +54,15 @@ const identifierTypeToElementType = (identifierInfo: FormulaIdentifierInfo): str
 
 const identifierTypeToElemIdType = (identifierInfo: FormulaIdentifierInfo): ElemIDType => (
```
[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/src/filters/formula_deps.ts:55]: I kind of like the separation to different layers here, where the parsing layer does its thing without bothering with Salto semantics

File: packages/salesforce-adapter/src/filters/formula_utils/parse.ts [hunk 3]
```diff
@@ -106,22 +106,85 @@ export const parseObjectType = (value: string): FormulaIdentifierInfo[] => {
...
 }
 
-export const parseFormulaIdentifier = (variableName: string, originalObject: string): FormulaIdentifierInfo[] => {
-  const types: FormulaIdentifierInfo[] = []
+const parseFieldIdentifier = (fieldWithPrefix: string, parentObject: string): FormulaIdentifierInfo[] => {
+  const field = _.trimStart(fieldWithPrefix, '$')
+
+  // Either the field identifier has an explicit parent (e.g. 'Account.Industry') or it implicitly refers to
+  // the provided parent object (e.g. 'Name')
+  const fieldParent = field.includes('.') ? getObject(field) : parentObject
```
[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/src/filters/formula_utils/parse.ts:116]: Makes sense to me.

**Pull request merged**